flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.metrics.prometheus;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.metrics.Metric;
+import org.apache.flink.metrics.MetricConfig;
+import org.apache.flink.metrics.reporter.MetricReporter;
+import org.apache.flink.metrics.reporter.Scheduled;
+import org.apache.flink.util.AbstractID;
+
+import io.prometheus.client.CollectorRegistry;
+import io.prometheus.client.exporter.PushGateway;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * /**
+ * {@link MetricReporter} that exports {@link Metric Metrics} via Prometheus Pushgateway.
+ */
+@PublicEvolving
+public class PrometheusPushGatewayReporter extends AbstractPrometheusReporter implements Scheduled {
+	private static final Logger LOG = LoggerFactory.getLogger(PrometheusPushGatewayReporter.class);
+
+	public static final String ARG_HOST = "host";
+	public static final String ARG_PORT = "port";
+
+	public static final char JOB_NAME_SEPARATOR = '-';
+	public static final String JOB_NAME_PREFIX = "flink" + JOB_NAME_SEPARATOR;
+
+	private PushGateway pushGateway;
+	private final String jobName;
+
+	public PrometheusPushGatewayReporter() {
+		String random = new AbstractID().toString();
+		jobName = JOB_NAME_PREFIX + random;
```

Comment:
The jobname should be configurable

Reply:
if the jobname is configurable, it means each taskmanager may use same jobname. 
if so, the metrics of tm(A) may be covered with the metrics of tm(B), etc

### for example, tm1, tm2 push the metrics at the same time
```
<dependency>
    <groupId>io.prometheus</groupId>
    <artifactId>simpleclient</artifactId>
    <version>0.0.26</version>
</dependency>

<dependency>
    <groupId>io.prometheus</groupId>
    <artifactId>simpleclient_pushgateway</artifactId>
    <version>0.0.26</version>
</dependency>




CollectorRegistry registry = new CollectorRegistry();
String sameJobName = "flink-job";

// taskmanager A
Gauge tm1 = Gauge.build().name("flink_taskmanager_Status_JVM_CPU_Time").help("tm jvm cpu").register(registry);
tm1.set(41);

PushGateway pg1 = new PushGateway("localhost:9091");
pg1.push(registry, sameJobName);


// taskmanager B
registry.clear();
Gauge tm2 = Gauge.build().name("flink_taskmanager_Status_JVM_CPU_Time").help("tm jvm cpu").register(registry);
tm2.set(42);

PushGateway pg2 = new PushGateway("localhost:9091");
pg2.push(registry, sameJobName);

```

### result, the metrics of tmA is covered with tmB
![image](https://user-images.githubusercontent.com/20113411/40636865-269828ce-6334-11e8-92e7-b222e6cfe6c0.png)

Reply:
That can only happen if 2 metrics have the exact same name and set of labels. Due to how the reporter is implemented this generally cannot happen.

Reply:
Hi, I did an experment, one jm with two tm, then send metrics to gateway, here's result
![image](https://user-images.githubusercontent.com/20113411/40646016-a3b2f188-635a-11e8-9e07-283f167179e3.png)

Reply:
so, it's better to make the prefix of jobname to distinguish different clusters, and each flink clusters use the same prefix

Reply:
The PushGateway UI is irrelevant though, the Prometheus UI is what matters.

![capture](https://user-images.githubusercontent.com/5725237/40647873-4355e0de-632d-11e8-8fb1-9beb811410ad.PNG)

Reply:
wait a second...

Reply:
well, I forgot to remove the random suffix in my test. With the suffix removed indeed only metrics from one TM shows up.

I woudl consider this a bug in the pushgateway; isn't the whole point of labels to differentiate between different instances?

Reply:
yes

Reply:
To implement this we will have to change the design of the new reporter significantly. This will entail moving more logic from the abstract class to the concrete `PrometheusReporter` class as it will no longer be shared. Ideally the abstract class will only contain shared utility methods, and does not dictate the actual report/notification logic.

The pushgateway reporter will create a separate collector for each grouping key (i.e. variable names/values, retrieved via `MetricGroup#getAllVariables()`). The collectors are not registered with the registry as this is only necessary in the pull model. The grouping by labels should ensure that we don't send to many requests; we may want to introduce a delay parameter to be safe.
A new metric will be assigned to a collector based on the hash of the variables map; we may not use the variables map or MetricGroup direclty as a mapping key as the map may change over time.
Similar to the existing vanilla reporter we will have to rely on ref-counting to cleanup existing collectors.

This will allow users to freely switch between both reporters without having to worry about any additional detail but the port configuration.

Reply:
What would happen if every taskmanager/jobmanager has it's own pushgateway?

Reply:
eh forget it, that's not a viable option for containerized environments that this issue targets anyway...

Reply:
:+1: right
it's better to use JobManager ID / TaskManager ID to compose the jobName, then jobName is `JM ID` / `TM ID`  
or combined with the specified prefix like `prefix + JM ID` / `prefix + TM ID`

but for now,  JM IDs are currently not exposed, so use random strings instead of JM/TM ID

Reply:
![image](https://user-images.githubusercontent.com/20113411/40701111-3b47089a-640f-11e8-8cd3-bc5684c07228.png)

Reply:
@zentol, please cc, and what else do I need to do?  :+1:

Reply:
I'm inclined to block the PR on the JobManager ID exposure. ([FLINK-9543](https://issues.apache.org/jira/browse/FLINK-9543))

The PR is not at risk at becoming outdated, so keeping it open for a while isn't a problem.

Reply:
ok，I see

Reply:
I just don't see a reason to rush this. There's a known issue we have to fix and the PR is not at risk of becoming outdated in the mean-time. If the argument is that "other people might start using it already" then we may just end up unnecessarily breaking their setup before the release.

Reply:
At the _very least_ we should already be using TM IDs.

Reply:
I'm not sure whether changing how the IDs are generated will break anyone's setup. My worry is just that this PR might fall through the cracks even after FLINK-9543 is being fixed. That would be a pity since it is a really nice feature from which users benefit.

Agreed that we should use the TM IDs for the TaskManager metrics.